bpf: don't look up trace ID if packet tracing via IP options is disabled

tklauser opened the pull request:
Currently, `check_and_store_ip_trace_id` incurs a map lookup on every call, even if packet tracing via IP options is disabled (`--ip-tracing-option-type=0`). Avoid this unnecessary lookup in case the feature is disabled.

Fixes: a860170c107d ("feat: add IPv4 packet tracing via IP options")

/cc @Bigdelle

Bigdelle commented:
This makes a lot of sense to me and thanks for catching this. I'm worried about a specific edgecase where there is a lingering trace ID stored in the map that never gets cleared when this feature gets disabled.

tklauser commented:
> This makes a lot of sense to me and thanks for catching this! I'm worried about a specific edgecase where there is a lingering trace ID stored in the map that never gets cleared when this feature gets disabled. Thus, this may lead to an extraneous trace ID for a non-tagged flow when we call `load_ip_trace_id()`, even when this feature is disabled.

I see, thanks for the explanation. To cover that case, maybe we should better check `CONFIG(tracing_ip_option_type) == 0` in `load_ip_trace_id` as well and `return 0` in case the feature is disabled. This also has the nice side effect that an additional, unconditional map lookup in `load_ip_trace_id` is avoided in case the feature is disabled. Or do you see any problem with that?

tklauser committed:
bpf: don't look up trace ID in map if tracing via IP options is disabled

Currently, check_and_store_ip_trace_id and load_ip_trace_id both incur a
map lookup on every call, even if packet tracing via IP options is
disabled (--ip-tracing-option-type=0). Avoid these unnecessary lookups
in case the feature is disabled.

Fixes: a860170c107d ("feat: add IPv4 packet tracing via IP options")
Signed-off-by: Tobias Klauser <tobias@cilium.io>
